### PR TITLE
[Status] Add absl::Status and absl::StatusOr

### PR DESCRIPTION
**Pull request by natsunoyoru97**
### Why we have the PR

- Import `Status` and `StatusOr` class to propogate errors.

### What this PR did

- Import `absl::Status` and `absl::StatusOr` in abseil
- Add script for compiling abseil in c++17 standard to avoid [issues triggered by pre-compiled abseil](https://github.com/abseil/abseil-cpp/blob/master/FAQ.md#what-is-abi-and-why-dont-you-recommend-using-a-pre-compiled-version-of-abseil). Pre-compiled abseil in xrepo does not cover all the testcases provided by abseil and fails to link to some functions.
- Enforce the project uses clang to compile (this should be done in #8, but bugs were found)

### Description of the bug
There is a bug like below if you use abseil in `xrepo`:
```
[ 75%]: linking.release test
...
error: /usr/bin/ld: /home/natsunoyoru/.xmake/packages/a/abseil/20220623.0/c60e4aa2b7614ef29c9a1e17127fd352/lib/libabsl_cordz_info.a(cordz_info.cc.o): in function `absl::lts_20220623::cord_internal::CordzInfo::~CordzInfo()':
cordz_info.cc:(.text+0x20): undefined reference to `absl::lts_20220623::Mutex::~Mutex()'
/usr/bin/ld: /home/natsunoyoru/.xmake/packages/a/abseil/20220623.0/c60e4aa2b7614ef29c9a1e17127fd352/lib/libabsl_cordz_info.a(cordz_info.cc.o): in function `absl::lts_20220623::cord_internal::CordzInfo::CordzInfo(absl::lts_20220623::cord_internal::CordRep*, absl::lts_20220623::cord_internal::CordzInfo const*, absl::lts_20220623::cord_internal::CordzUpdateTracker::MethodIdentifier)':
cordz_info.cc:(.text+0x7d3): undefined reference to `absl::lts_20220623::Now()'
/usr/bin/ld: /home/natsunoyoru/.xmake/packages/a/abseil/20220623.0/c60e4aa2b7614ef29c9a1e17127fd352/lib/libabsl_cordz_info.a(cordz_info.cc.o): in function `absl::lts_20220623::cord_internal::CordzInfo::Untrack()':
cordz_info.cc:(.text+0x9c8): undefined reference to `absl::lts_20220623::Mutex::Lock()'
/usr/bin/ld: cordz_info.cc:(.text+0x9de): undefined reference to `absl::lts_20220623::Mutex::Unlock()'
/usr/bin/ld: /home/natsunoyoru/.xmake/packages/a/abseil/20220623.0/c60e4aa2b7614ef29c9a1e17127fd352/lib/libabsl_cordz_info.a(cordz_info.cc.o): in function `absl::lts_20220623::cord_internal::CordzInfo::Lock(absl::lts_20220623::cord_internal::CordzUpdateTracker::MethodIdentifier)':
cordz_info.cc:(.text+0xb11): undefined reference to `absl::lts_20220623::Mutex::Lock()'
...
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```
### Steps to explore test this bug
- Switch to `natsu/experimental` branch and revert to the commit with the comment "exploring test - abseil package" if needed
- Remove abseil if you got one:
```shell
xrepo remove abseil
```
- Clean all the binary files
```shell
xmake c
```
- Reinstall the abseil package
```shell
xmake f -c
```
- Build the project in verbose
```shell
xmake -v
```

### Testcase
The testcases are integrated in `test/main.cc` and you can run `xmake run test` to test.
```c++
absl::Status IsBoolTrue(bool b) {
  return (b) ? absl::OkStatus() : absl::InvalidArgumentError("b is not true.");
}

absl::StatusOr<std::string> IsStringAlpha(std::string&& s) {
  return std::all_of(s.cbegin(), s.cend(), isalpha)
             ? absl::StatusOr<std::string>(s)
             : absl::StatusOr<std::string>(absl::InvalidArgumentError(
                   "the string has non-alpha characters."));
}

TEST(UtilsTest, StatusWorks) {
  EXPECT_EQ(IsBoolTrue(true).ok(), true);
  EXPECT_NE(IsBoolTrue(false).ok(), true);
}

TEST(UtilsTest, StatusOrWorks) {
  EXPECT_EQ(IsStringAlpha("abc").ok(), true);
  EXPECT_NE(IsStringAlpha("abc1").ok(), true);
}
```

### CODE REVIEW

File: test/main.cc [hunk 1]
```diff
@@ -1,11 +1,40 @@
...
+
+absl::StatusOr<std::string> IsStringAlpha(std::string&& s) {
+  return std::all_of(s.cbegin(), s.cend(), isalpha)
+             ? absl::StatusOr<std::string>(s)
+             : absl::StatusOr<std::string>(absl::InvalidArgumentError(
+                   "the string has non-alpha characters."));
+}
+
+TEST(UtilsTest, StatusWorks) {
+  EXPECT_EQ(IsBoolTrue(true).ok(), true);
```
Comment (dentiny) [inline test/main.cc:29]: `EXPECT_TRUE` and `EXPECT_FALSE`

File: test/main.cc [hunk 2]
```diff
@@ -1,11 +1,40 @@
...
 TEST(DatabaseTest, EchoWorks) { EXPECT_STRNE("hello", "world"); }
 
 TEST(DatabaseTest, AddWorks) { EXPECT_STREQ("hello", "hello"); }
 
+// Test for Utils
+absl::Status IsBoolTrue(bool b) {
+  return (b) ? absl::OkStatus() : absl::InvalidArgumentError("b is not true.");
+}
+
+absl::StatusOr<std::string> IsStringAlpha(std::string&& s) {
```
Comment (dentiny) [inline test/main.cc:21]: What's the purpose for `&&`....

File: xmake.lua [hunk 3]
```diff
@@ -77,28 +77,73 @@ DB_LLVM_TEST_FLAGS = {
...
 
-set_languages("c++17")
-set_toolset("cxx", "clang++")
-
-add_requires("abseil", "doctest", "gtest ~1.12.1", "glog", "gflags")
+set_languages("cxx17")
+add_requires("doctest", "gtest", "glog", "gflags")
 add_rules("mode.debug", "mode.release")
 
+-- Pre-compiled Abseil is NOT PREFERRED, see https://github.com/abseil/abseil-cpp/blob/master/FAQ.md#what-is-abi-and-why-dont-you-recommend-using-a-pre-compiled-version-of-abseil.
```
Comment (dentiny) [inline xmake.lua:84]: I don't understand why we would have ABI issue here?

File: xmake.lua [hunk 4]
```diff
@@ -77,28 +77,73 @@ DB_LLVM_TEST_FLAGS = {
...
+-- Abseil should be built with the flag "-std=c++17", see https://github.com/abseil/abseil-cpp/issues/819.
+package("abseil")
```
Comment (dentiny) [inline xmake.lua:86]: Curious what's the difference between separately declaring a "abseil" package VS "add_requires", like what you do for gtest?

**Comment by dentiny** [comment]
Want to follow up on this PR?

**Review by dentiny — APPROVED** [review]
Mind leaving a TODO on abseil check-in? I still think it could be resolved without manually check-in all symbols.

**Comment by natsunoyoru97** [comment]
> Mind leaving a TODO on abseil check-in? I still think it could be resolved without manually check-in all symbols.

Cool. I will have a try for a better solution.

**Pull request merged**